Data snapshot query execution

MishaDemianenko opened the pull request:
Introduce possibility to execute query based on data snapshot.
Any data modification will mark corresponding pages with a version that
will be equal to transaction id in which that particular change was introduced.
Also allow count store to track transactions in which changes where introduced.

To be able to guarantee that query result is based on a data that was
present on query execution start and was not modified while it was running,
the engine will verify that all paged that were accessed during query
execution have the version that is less or equal that was last closed
on a moment when query started.

Support for snapshot query execution in CC and HA

Snapshot query execution is enabled by 'unsupported.dbms.query.snapshot'
setting and number of retries that engine will do while trying to get
stable snapshot is controlled by 'unsupported.dbms.query.snapshot.retries'.

MishaDemianenko committed:
Data snapshot query execution

Introduce possibility to execute query based on data snapshot.
Any data modification will mark corresponding pages with a version that
will be equal to transaction id in which that particular change was introduced.
Also allow count store to track transactions in which changes where introduced.

To be able to guarantee that query result is based on a data that was
present on query execution start and was not modified while it was running,
the engine will verify that all paged that were accessed during query
execution have the version that is less or equal that was last closed
on a moment when query started.

Support for snapshot query execution in CC and HA

Snapshot query execution is enabled by 'unsupported.dbms.query.snapshot'
setting and number of retries that engine will do while trying to get
stable snapshot is controlled by 'unsupported.dbms.query.snapshot.retries'.

MishaDemianenko committed:
Do not clear context when statement is closed

MishaDemianenko committed:
Allow tracking version context on transaction that executed by multiple
threads (transactions that goes over rest end point).
Re-check version context when cursor repositioned to its current page again.

Check version context state after eager result visiting during serialization
since both bolt and rest writers will try to load all properties of
nodes and relationships from database, if those are part of result,
to send those to client as well. And since those executed outside of query
we need to expand version check to verify that after streaming result
we still based on valid snapshot.
If dirty context will be noticed exception will be thrown since we can't
restart that query - its too late and client already get partial results for it.

MishaDemianenko committed:
Formatting and naming comments.

MishaDemianenko committed:
Update new fields access to use unsafe only, tests for newly
introduced helper methods

MishaDemianenko committed:
Use correct volatile long get instead of simple long get

MishaDemianenko committed:
Update tracking of highest page file version.

Make timeout for CC test a bit longer.

MishaDemianenko committed:
Give a bit more time for clustering test

MishaDemianenko committed:
Fix compilation errors

MishaDemianenko committed:
Initial version of page version tracking where page is 64 bytes
and version tracking tests working.

MishaDemianenko committed:
Perform re-execution of the query also when paremeters passed as MapValue.

MishaDemianenko committed:
Make swapper id short, update mapping to use only 2 bytes to map swapper id.

MishaDemianenko committed:
Combine page id, swapper id and usage stamp into single one single long.
Make page list record 32 bytes again.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/context/EmptyVersionContextSupplier.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class EmptyVersionContextSupplier implements VersionContextSupplier
+{
+    public static final VersionContextSupplier INSTANCE = new EmptyVersionContextSupplier();
```
If this was an `EMPTY` or `NULL` constant on the `VersionContextSupplier` it'd be easier to find and this class wouldn't have to be public. Small thing.

chrisvest replied:
Similar with the `EmptyVersionContext`.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/context/VersionContext.java:
```diff
@@ -0,0 +1,72 @@
+ * Write context data modification with have version that is equal to transaction that is currently committing.
+ *
+ * As soon reader that associated with a context will observe data version that it should not see, context will be
+ * marked as dirty.
```
This comment is not easy to follow. Could you tighten up the wording, and add something about the life cycle of these instances, and the temporal coupling of their methods?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsTracker.java:
```diff
@@ -293,6 +264,43 @@ protected void writeFormatSpecifier( WritableBuffer formatSpecifier )
         formatSpecifier.put( 0, FORMAT );
     }
 
+    private static class CountsTrackerRotationMonitor implements RotationMonitor
```
👍

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/io/pagecache/impl/muninn/VersionContextTrackingIT.java:
```diff
@@ -0,0 +1,214 @@
+//                    {
+//                        maxTransactionId = Math.max( maxTransactionId, pageCursor.lastTxModifierId() );
+//                    }
+//                }
```
This looks important?

chrisvest commented on enterprise/causal-clustering/src/test/java/org/neo4j/io/pagecache/impl/muninn/VersionContextTrackingIT.java:
```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+//    private class CursorPageAccessor extends MuninnPageCursor
```
Missing cleanup?

chrisvest commented on enterprise/ha/src/test/java/org/neo4j/kernel/ha/UpdatePullerTriggersPageTransactionTrackingIT.java:
```diff
@@ -0,0 +1,224 @@
+            assertEquals( "Unable to get clean data snapshot for query 'match (n) return n' after 5 attempts.", executionException.getMessage());
+        }
+        byzantineIdSupplier.useCorrectTxId();
+        slave.execute( "match (n) return n" ).close();
```
Where does the update puller get involved in this test?

MishaDemianenko replied:
here we check that update pulling together with starting with tx on slave works as expected:
we check first that in case if our incorrect tx id provider will supply wrong tx id we will not gonna be able to execute query and in this line we check that as soon as id provider supply correct transaction id query execution works as expected

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -119,12 +120,27 @@ public final boolean next( long pageId ) throws IOException
+    void verifyContext()
+    {
+        VersionContext versionContext = versionContextSupplier.getVersionContext();
+        if ( versionContext.lastClosedTransactionId() == Long.MAX_VALUE )
```
The interface should document that this value is special, then.

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java:
```diff
@@ -408,7 +413,9 @@ public void markContextAsDirtyWhenAnyEvictedPageHaveModificationTransactionHighe
-        pageCache.evictPages( 1, 1, EvictionRunEvent.NULL );
+
+        pageCache.pages.tryEvict( pageCache.grabFreeAndExclusivelyLockedPage( PageFaultEvent.NULL ),
+                EvictionRunEvent.NULL );
```
It doesn't make sense to try to evict a free page. The `tryEvict` will always return `false` because the given page is already exclusively locked. The `grabFreeAndExclusivelyLockedPage` method might do a cooperative eviction, but the evicted page will be chosen at random. Can we make this test deterministic?

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java:
```diff
@@ -441,7 +448,9 @@ public void doNotMarkContextAsDirtyWhenAnyEvictedPageHaveModificationTransaction
-        pageCache.evictPages( 1, 1, EvictionRunEvent.NULL );
+
+        pageCache.pages.tryEvict( pageCache.grabFreeAndExclusivelyLockedPage( PageFaultEvent.NULL ),
+                EvictionRunEvent.NULL );
```
Same as above.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -363,12 +362,22 @@ public boolean decrementUsage( long pageRef )
-        return UnsafeUtil.getLong( offFilePageId( pageRef ) );
+        int highByte = UnsafeUtil.getByte( offFilePageId( pageRef ) ) & UNSIGNED_BYTE_MASK;
+        long lowInt = UnsafeUtil.getInt( offFilePageId( pageRef ) + Byte.BYTES ) & UNSIGNED_INT_MASK;
+        long filePageId = (((long) highByte) << Integer.SIZE) | lowInt;
```
Can't you replace this with `UnsafeUtil.getLong( offFilePageId( pageRef ) ) >>> 24`? This method is called as part of `isBoundTo`, which in turn is called on every pin.

Speaking of `isBoundTo`, the file page id and the swapper id are now part of the same aligned word. So could check the binding with a single compare by masking/shifting out the usage stamp, and comparing it with a reconstructed binding word. A single compare for both fields.

MishaDemianenko replied:
talked offline about it, we will look on it separately to see if it make sense to do it

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/PageList.java:
```diff
@@ -42,28 +42,31 @@
+    private static final long UNSIGNED_INT_MASK = 0xFFFFFFFFL;
+
+    // 40 bits for file page id
+    private static final long MAX_FILE_PAGE_ID = 0b11111111_11111111_11111111_11111111_11111111L;
```
👍 8 pebibytes should be enough for anyone.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/SwapperSet.java:
```diff
@@ -98,8 +101,7 @@ synchronized short allocate( PageSwapper swapper )
             {
-                //TODO safe cast
-                short id = (short) free.iterator().next();
+                short id = Numbers.safeCastIntToShort( free.iterator().next() );
```
You already got this in a static import.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCursor.java:
```diff
@@ -116,12 +120,41 @@ public final boolean next( long pageId ) throws IOException
+    {
+        VersionContext versionContext = versionContextSupplier.getVersionContext();
+        long lastClosedTransactionId = versionContext.lastClosedTransactionId();
+        if ( lastClosedTransactionId == Long.MAX_VALUE )
```
Resurfacing comment: The interface should document that this value is special, then.

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java:
```diff
@@ -214,6 +220,247 @@ public void mustFlushDirtyPagesOnEvictingAllPages() throws Exception
+        }
+
+        pageCache.pages.tryEvict( pageCache.grabFreeAndExclusivelyLockedPage( PageFaultEvent.NULL ),
+                EvictionRunEvent.NULL );
```
Resurfacing comment: It doesn't make sense to try to evict a free page. The `tryEvict` will always return `false` because the given page is already exclusively locked. The `grabFreeAndExclusivelyLockedPage` method might do a cooperative eviction, but the evicted page will be chosen at random. Can we make this test deterministic?

chrisvest commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCacheTest.java:
```diff
@@ -214,6 +220,247 @@ public void mustFlushDirtyPagesOnEvictingAllPages() throws Exception
+        }
+
+        pageCache.pages.tryEvict( pageCache.grabFreeAndExclusivelyLockedPage( PageFaultEvent.NULL ),
+                EvictionRunEvent.NULL );
```
Resurfacing comment: Same as above.

chrisvest reviewed (changes requested):
Looks good overall, but there's still some cleanup and test fixing left to do.

MishaDemianenko committed:
Review comments:
rename empty instances, update page cache eviction tests,
fix eviction boundaries check, cleanup CC version tracker test.

MishaDemianenko committed:
Review comments: better javadoc.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/tracing/cursor/context/VersionContext.java:
```diff
@@ -22,11 +22,24 @@
+ *     some version that write context was initialised with.
+ *     <br/>
+ *     For example, commit will start write context that with a version that is equal to current
+ *     committing transaction id./li>
```
Missed a `<` on the li-end tag, but otherwise 👍

MishaDemianenko committed:
Review comments: even better javadoc.

MishaDemianenko commented:
@chrisvest now it should be it, updated

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/SwapperSet.java:
```diff
@@ -107,7 +110,7 @@ synchronized int allocate( PageSwapper swapper )
 
         // No free slot was found above, so we extend the array to make room for a new slot.
-        int id = swapperMappings.length;
+        short id = safeCastLongToShort( swapperMappings.length );
```
I guess this should say `safeCastIntToShort` instead of `Long`?

MishaDemianenko committed:
Use proper safe cast
